### PR TITLE
Adjust the way parser handles line-If termination.

### LINKED ISSUE #14761: Missing parser error BC30205: End of statement expected.
```
Module Module1

    Sub Main()

    End Sub

    Sub Test(val As Object)
        Dim r As String

        If val Is Nothing Then r = "null" GoTo label1

label1:
    End Sub
End Module
```

Expected:

```
(11) : error BC30205: End of statement expected.

        If val Is Nothing Then r = "null" GoTo label1
                                          ~~~~~~~~~~~
```

Observed: No error, compilation succeeds.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #14761.

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Parser/BlockContexts/SingleLineElseContext.vb [hunk 1]
```diff
@@ -92,34 +93,90 @@ EndBlock:
                     End If
                     GoTo EndBlock
 
+                Case Else
+                    Throw ExceptionUtilities.UnexpectedValue(token.Kind)
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Parser/BlockContexts/SingleLineElseContext.vb]: Am I right in assuming this fail safe would've caught this case and prevented a successful compilation?

File: src/Compilers/VisualBasic/Portable/Parser/BlockContexts/SingleLineElseContext.vb [hunk 2]
```diff
@@ -92,34 +93,90 @@ EndBlock:
...
-        Friend Overrides Function ProcessOtherAsStatementTerminator() As BlockContext
-            Dim context = EndBlock(Nothing)
-            Return context.ProcessOtherAsStatementTerminator()
+        Private ReadOnly Property TreatElseAsStatementTerminator As Boolean
+            Get
+                ' We can treat 'Else' as a valid statement terminator only
+                ' if there is a line-If without Else up the block context chain, or
+                ' if we are inside a single-line statement lambda and the 'Else' terminates the lambda.
+                Debug.Assert(BlockKind = SyntaxKind.SingleLineElseClause)
+                Debug.Assert(PrevBlock.BlockKind = SyntaxKind.SingleLineIfStatement)
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Parser/BlockContexts/SingleLineElseContext.vb]: It feels wrong to have a property which can only be contextually invoked.

File: src/Compilers/VisualBasic/Portable/Parser/BlockContexts/SingleLineElseContext.vb (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Parser/BlockContexts/SingleLineElseContext.vb]: Yes, before the change we were allowing all kinds of things to pass through the Case Else.

File: src/Compilers/VisualBasic/Portable/Parser/BlockContexts/SingleLineElseContext.vb (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Parser/BlockContexts/SingleLineElseContext.vb]: The property can always be invoked, this asserts an invariant in the way block contexts are nested for line-if. They are always nested this way.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

**Comment by AdamSpeight2008** [comment]
@AlekseyTs  Could you provide some short code examples of what was allowed and now isn't?

**Comment by AlekseyTs** [comment]
> Could you provide some short code examples of what was allowed and now isn't?

The issue referenced at the top, #14761, has short example. I also added a bunch of unit-tests, you can compare compiler's behavior for them for before and after the change.

**Comment by cston** [comment]
LGTM

**Pull request merged**